CoreFeatures/HttpEndpoint/DeserializeStrategyPostMultipartFormData/src/main/java/info/smart_tools/smartactors/http_endpoint/deserialize_strategy_post_multipart_form_data/DeserializeStrategyPostMultipartFormData.java

```diff
@@ -0,0 +1,208 @@
+                    );
+                    environment.setValue(messageFN, message);
+                    environment.setValue(dataFN, data);
+                    environment.setValue(uploadDirectoryFN, uploadDirectory);
+                    action.execute(environment);
+                } catch (Exception e) {
+                    throw new RuntimeException("Failed to deserialize request", e);
+                }
+            });
+
+            return message;
+        } catch (Exception e) {
+            throw new DeserializationException("Failed to deserialize request", e);
+        }
+    }
+
+    private void initializeActions() {
+        this.actions.put(
+                HttpDataType.Attribute, (IObject env) -> {
+                    try {
+                        IObject message = (IObject) env.getValue(messageFN);
+                        InterfaceHttpData data = (InterfaceHttpData) env.getValue(dataFN);
+                        IFieldName fieldName = IOC.resolve(
+                                Keys.getKeyByName(IFIELD_NAME_STRATEGY_NAME), data.getName()
+                        );
+                        message.setValue(fieldName, ((Attribute) data).getValue());
+                    } catch (ResolutionException e) {
+                        throw new RuntimeException(
+                                "Could not resolve field name by key name - 'info.smart_tools.smartactors.iobject.ifield_name.IFieldName'",
+                                e
+                        );
+                    } catch (IOException e) {
+                        throw new RuntimeException(
+                                "Could not read value from HTTP request by given attribute name",
+                                e
+                        );
+                    } catch (ChangeValueException | InvalidArgumentException e) {
+                        throw new RuntimeException(
+                                "Could not set attribute value to the message",
+                                e
+                        );
+                    } catch (ReadValueException e) {
+                        throw new RuntimeException(
+                                "Could not read value from environment.",
+                                e
+                        );
+                    }
+                }
+        );
+        this.actions.put(
+            HttpDataType.FileUpload, (IObject env) -> {
+                try {
+                    IFieldName fieldName = IOC.resolve(
+                            Keys.getKeyByName(IFIELD_NAME_STRATEGY_NAME),
+                            FILENAME_FIELD_NAME
+                    );
+                    IObject message = (IObject) env.getValue(messageFN);
+                    FileUpload uploadedFile = (FileUpload) env.getValue(dataFN);
+                    String uploadDir = (String) env.getValue(uploadDirectoryFN);
+                    final File file = new File(uploadDir + File.separator + uploadedFile.getName());
```

Comment:
It's possible to go out of the upload directory if the file name is like "../../filename".
Need to check that the resulting absolute file is always in the upload dir.

Reply:
fixed.